Reservable items (rooms and washing machines)

viktorcsimma opened the pull request:
As in the title. It supports:
- washing machines (only one-hour slots, and at most 6 of them),
- rooms (with arbitrary intervals, even recurring, but they need to be verified by someone authorized).

Known issues:
- ~~The timetable sometimes disappears when stepping forwards and then backwards.~~
- Reservable items cannot be created from the UI yet, only in the database.
- The code structure could be better. I have considered putting everything under the `Dormitory` folder so as not to pollute the upper folders.
- ~~I have not written test cases yet; I would like to know what you think before proceeding with that.~~

Thanks to everyone at the course for helping:)

viktorcsimma committed:
Progress of 21st March

viktorcsimma committed:
Progress on 11th April

viktorcsimma committed:
Progress on 18th April

viktorcsimma committed:
Progress on 25th April

viktorcsimma committed:
Progress on 2nd May

viktorcsimma committed:
Progress on 9th May

viktorcsimma committed:
Progress on 16th May

viktorcsimma committed:
Progress on 21st May

viktorcsimma committed:
Merge development

viktorcsimma committed:
Handling out-of-order items with a boolean flag

viktorcsimma committed:
Starting to build a user-friendly interface; adjustments for washing machines

viktorcsimma committed:
Fixing conflict inspection

viktorcsimma committed:
Splitting free blocks at every hour

viktorcsimma committed:
Implementing additional restrictions for washing machines; removing auto-verification

viktorcsimma committed:
Fixing breadcrumb for washing machines

viktorcsimma committed:
Handling out-of-order items; other small fixes

viktorcsimma committed:
Added fault reporting and toggling out_of_order flag

viktorcsimma committed:
Merge branch 'development' into reservable-items-final

viktorcsimma committed:
Fixing seeder so that it only uses verified users

viktorcsimma committed:
Fixing and enhancing verification logic

viktorcsimma committed:
Styling buttons at the bottom of 'show' cards

viktorcsimma committed:
Fixing an illegal conversion when creating a recurring event

viktorcsimma committed:
Localization

viktorcsimma committed:
More localization

viktorcsimma committed:
Removing unnecessary columns

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in c69c5ae according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/623

kdmnk commented on app/Http/Controllers/ReservableItemController.php:
```diff
@@ -0,0 +1,129 @@
+        } elseif ('washing_machine' == $type) {
+            $items = ReservableItem::where('type', 'washing_machine')->get();
+        } elseif ('room' == $type) {
+            $items = ReservableItem::where('type', 'room')->get();
```
Make sure to validate it, then you can use:

$items = ReservableItem::query().
If ($request->type){
$items->where('type', $request->type)
}
$items->get()

kdmnk commented on app/Http/Controllers/ReservableItemController.php:
```diff
@@ -0,0 +1,129 @@
+    {
+        $this->authorize('viewAny', ReservableItem::class);
+
+        $from = Carbon::today()->startOfWeek();
+        $until = $from->copy()->addDays(6);
```
DIdn't you want to add this as part of the request?

viktorcsimma replied:
Actually, they aren't needed anymore; the Livewire element handles that.

kdmnk commented on app/Http/Controllers/ReservableItemController.php:
```diff
@@ -0,0 +1,129 @@
+
+        $validator = Validator::make($request->all(), [
+            'name' => 'required|string|max:255',
+            'type' => Rule::in(["washing_machine", "room"]),
```
Store these as const in the model. Or you can already start using https://github.com/EotvosCollegium/mars/issues/388

viktorcsimma replied:
Good idea; I'll use a backed enum.

kdmnk commented on app/Http/Controllers/ReservableItemController.php:
```diff
@@ -0,0 +1,129 @@
+
+        $outOfOrder = !($item->out_of_order);
+        $item->out_of_order = $outOfOrder;
+        $item->save();
```
You can make it a one liner:
$item->update(['out_of_order' => !$item->out_of_order])

kdmnk commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,488 @@
+    /**
+     * Lists reservations for all washing machines.
+     */
+    public function indexForWashingMachines()
```
This should be in the reservableItemController. In fact, the index there already does this without the dates (which I don't see why do you need here?).

viktorcsimma replied:
I wanted it to resemble the real washing machine reservation book at the reception; this was the easiest way to manage that. Or do you have another idea?

viktorcsimma replied:
Now, I could remove that separate route and solve it via the index route.

kdmnk commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,488 @@
+    private static function reachedMaximumForWashingMachines(User $user): bool
+    {
+        return ReservableItem::MAX_WASHING_RESERVATIONS
+          <= Reservation::where('user_id', $user->id)
+                ->where('reserved_until', '>', Carbon::now())
+                ->whereExists(function (Builder $query) {
+                    $query->select(DB::raw(1))
+                          ->from('reservable_items')
+                          ->where('type', ReservableItem::WASHING_MACHINE)
+                          ->where('out_of_order', false)
+                          ->whereColumn('reservable_items.id', 'reservations.reservable_item_id');
+                })->count();
```
What are you trying to check here exactly?
In my intuition, this should be that a user can't have more than x reservations in the future but I see that it's not connected to the user. 

I would make it something like this but I realised you might want something else:

$user->reservedItems(ReservableItem::WASHING_MACHINE)->count() >= ReservableItem::MAX_WASHING_RESERVATIONS

If you can make this as simple as this maybe we don't even need this separated as a private function.

viktorcsimma replied:
The idea is that they should not reserve more than 6 washing slots at once. So I count the number of reservations that belong to them, are for washing machines and end in the future.

kdmnk commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,488 @@
+     */
+    private static function hasConflict(Reservation $newReservation): ?Reservation
+    {
+        $conflictingReservations = ReservableItem::find($newReservation->reservable_item_id)
```
Use relation. $newReservation->reservableItem->...

viktorcsimma replied:
True; I don't know why I have done it so:D

kdmnk commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,488 @@
+                }
+                if (Carbon::make($request->reserved_until) < Carbon::now()) {
+                    $validator->errors()->add('reserved_until', __('reservations.in_past'));
+                }
```
you can use rules like 'date|after:reserved_from|after:now'

kdmnk commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,488 @@
+                    $validator->errors()->add('reserved_until', __('reservations.in_past'));
+                }
+                if (isset($request->recurring)
+                           && Carbon::make($request->reserved_from)
+                             > Carbon::make($request->last_day)->addDay()) {
+                    $validator->errors()->add(
+                        'last_day',
+                        __('reservations.last_day_before_first')
+                    );
```
I would add a new validation rule for this (https://laravel.com/docs/11.x/validation#custom-validation-rules)

just because this function is long already

viktorcsimma replied:
Huh; wouldn't it just create more unnecessary code to create an entire new class for a rule only used once? But I'm trying to shorten the code with `sometimes` rules.

kdmnk commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,488 @@
+            });
+        } else {
+            $validator->after(function ($validator) use ($request, $item) {
+                $from = Carbon::make($request->reserved_from);
+                // we only allow one-hour slots for washing machines
+                if (0 != $from->minute || 0 != $from->second) {
+                    $validator->errors()->add('reserved_from', __('reservations.one_hour_slot_only'));
+                }
```
why don't you just create a new date at from xx:00:00 for a given xx:yy:zz. 
If it is clear in the UI then the backend can just ignore invalid times.

viktorcsimma replied:
Well, it is not yet clear in the UI :sweat_smile: I should probably use a simple integer input there for the hour.

kdmnk commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,488 @@
+                if (0 != $from->minute || 0 != $from->second) {
+                    $validator->errors()->add('reserved_from', __('reservations.one_hour_slot_only'));
+                }
+                if (Carbon::make($request->reserved_from)->addHour() < Carbon::now()) {
```
you can still use the built in after/before rules. Pass a string that represents Carbon::now()->subHour()

kdmnk commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,488 @@
+        if (isset($validatedData['recurring']) && $item->isWashingMachine()) {
+            return redirect()->back()
+                ->with('error', __('reservations.no_recurring_for_washing_machines'));
+        }
```
just check and ignore the value when saving (to make the function simpler). Again, I assume the UI does not allow this.

kdmnk commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,488 @@
+        ]);
+    }
+
+    public const THIS_ONLY = 'this_only';
```
prefix these with EDIT_

kdmnk commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,488 @@
+                    );
+                }
+            });
+        }
```
same here, try to eliminate ->after() cases

kdmnk commented on app/Livewire/Timetable.php:
```diff
@@ -0,0 +1,171 @@
+            }
+        }
+
+        return $splitBlocks;
```
how could we do that?

viktorcsimma replied:
I've now put the splitting logic into a different function.

kdmnk commented on app/Models/ReservationGroup.php:
```diff
@@ -0,0 +1,299 @@
+        'default_title',
+        'default_from',
+        'default_until',
+        'default_note',
```
I find the terminology default misleading. What about 'group_' as prefix?

kdmnk reviewed (changes requested):
TBH I gave up reviewing after the controllers.
I'll continue later but my general thoughts are that it's too complex.
I suggest to switch the development process for new modules in the future to aim for an MVP first. For example here, without approval, conflicts, recurring reservations. And then add these later once we know the basics are working. 
Since we are working online, in general, if you are unsure what's the best way to go feel free to open a pr and discuss it when questions arise. 

I will think how can we eliminate the checks to distinguish between washing machines and rooms every time.

And some thoughts about recurring events: 
Take a look at google calendar, they don't care about conflicts. I haven't gone through your code but seems like checking for conflicts is a pain in the ***. Maybe we could just ignore them and set it as a new conflicting state that needs human intervention for some occasions.
Another idea that we can do is to have a standard week schedule that is connected to the semester and then we have one time events for every individual week. 
But ofc if your solution works you can keep it now, just make sure to document what your functions are doing in detail.

Otherwise, I've tried the UI, it is pretty neat, the table looks good, nice job. For the rooms though, I was thinking would not it be better to show only one day once, with all the possible rooms as columns? More often the use case is to find any room at a given time and with the current solution you need to check each room individually.

viktorcsimma committed:
Applying smaller suggestions

viktorcsimma committed:
Fixing a test case

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 9eece32 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/623

viktorcsimma committed:
Fixing typos after search-and-replace

viktorcsimma committed:
Added documentation comments, fixed wrong prefixes

viktorcsimma committed:
Fixing equality checks on backed enums

viktorcsimma committed:
Applying phpstan suggestions

viktorcsimma committed:
Simplifying block logic in timetable component

viktorcsimma committed:
Improving reservation creation UI

viktorcsimma committed:
Adding timetable with all rooms to the room index page

viktorcsimma committed:
Simplifying Livewire object

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 61fe4e2 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/623

viktorcsimma committed:
Fixing disappearing blocks

viktorcsimma commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,454 @@
+
+                $verified = $reservation->group->verified
+                        && is_null($groupFrom) && is_null($groupUntil);
+
+                if (self::EDIT_ALL_AFTER == $validatedData['for_what']) {
+                    $reservation->group->setForAllAfter(
+                        firstReservation: $reservation,
+                        groupItem: $groupItem,
+                        user: $user,
+                        // ?int $frequency = null, // it cannot be set for now
+                        groupTitle: $groupTitle,
+                        groupFrom: $groupFrom,
+                        groupUntil: $groupUntil,
+                        groupNote: $groupNote,
+                        verified: $verified
+                    );
+                } else { // self::EDIT_ALL== $validatedData['for_what']
+                    $reservation->group->setForAll(
+                        groupItem: $groupItem,
+                        user: $user,
+                        // ?int $frequency = null, // it cannot be set for now
+                        groupTitle: $groupTitle,
+                        groupFrom: $groupFrom,
+                        groupUntil: $groupUntil,
+                        groupNote: $groupNote,
+                        verified: $verified
+                    );
+                }
+
+                // and for the last day:
+                if ($validatedData['last_day'] != $reservation->group->last_day) {
+                    $reservation->group->setLastDay($validatedData['last_day']);
+                }
+            } catch (ConflictException $e) {
+                abort(409, $e->getMessage());
+            }
+        }
+
+        return redirect()->route('reservations.show', $reservation);
+    }
```
I see... and can I include this besides an existing rule in the array?

viktorcsimma commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,454 @@
+    /**
+     * Checks whether the maximum number of reservations for washing machines has been reached.
+     */
+    private static function reachedMaximumForWashingMachines(User $user): bool
+    {
+        return ReservableItem::MAX_WASHING_RESERVATIONS
+          <= Reservation::where('user_id', $user->id)
+                ->where('reserved_until', '>', Carbon::now())
+                ->whereExists(function (Builder $query) {
+                    $query->select(DB::raw(1))
+                          ->from('reservable_items')
+                          ->where('type', ReservableItemType::WASHING_MACHINE)
+                          ->where('out_of_order', false)
+                          ->whereColumn('reservable_items.id', 'reservations.reservable_item_id');
+                })->count();
```
Nice:) thanks!

viktorcsimma commented on app/Models/Reservation.php:
```diff
@@ -0,0 +1,95 @@
+    /**
+     * A name to be displayed in a timetable.
+     */
+    public function displayName(): string
+    {
+        if ($this->reservableItem->isWashingMachine()) {
+            return $this->user->name;
+        } else {
+            return (
+                isset($this->title)
+                ? $this->title
+                : $this->user->name
+            ) . (
+                $this->verified
+                ? ''
+                : (' (' . __('reservations.unverified') . ')')
+            );
+        }
+    }
```
The user is always non-null; so it will be fine this way.

viktorcsimma commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,453 @@
+                if (0 != $from->minute || 0 != $from->second) {
+                    $validator->errors()->add('reserved_from', __('reservations.one_hour_slot_only'));
+                }
+            });
```
Done:)

viktorcsimma commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,453 @@
+
+        return ($conflictingReservations
+                    ->filter(fn ($reservation) => $reservation->id != $newReservation->id)
+                    ->first());
```
Done:)

viktorcsimma commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,453 @@
+            $reservation->reserved_from = Carbon::make($validatedData['reserved_from']);
+            $reservation->reserved_until =
+                $reservation->reservableItem->isWashingMachine()
+                ? $reservation->reserved_from->copy()->addHour()
```
Done:)

viktorcsimma replied:
I've forgot to rename the prefix; that was the problem.

viktorcsimma commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,453 @@
+
+        $reservation->group->delete();
+        return redirect()->route('reservations.items.show', $reservation->reservableItem);
+    }
```
Done:)

viktorcsimma commented on app/Models/ReservationGroup.php:
```diff
@@ -0,0 +1,299 @@
+                }
+            }
+        });
+    }
```
Done:)

viktorcsimma commented on app/Models/ReservationGroup.php:
```diff
@@ -0,0 +1,299 @@
+        }
+        $this->last_day = $newLastDay;
+        $this->save();
+    }
```
Done:) I've inserted a backslash.

viktorcsimma commented on app/Models/ReservationGroup.php:
```diff
@@ -0,0 +1,299 @@
+
+use App\Models\Reservation;
+
+class ConflictException extends \Exception
+{
+}
```
I'll do that.

viktorcsimma commented on database/migrations/2019_10_19_193746_create_workshops_table.php:
```diff
@@ -18,7 +18,7 @@ public function up()
             $table->text('name');
         });
-        foreach (Workshop::ALL as $key => $workshop) {
+        foreach (Workshop::ALLas $key => $workshop) {
```
Done:)

viktorcsimma commented on database/migrations/2019_10_13_130718_create_roles_table.php:
```diff
@@ -22,7 +22,7 @@ public function up()
 
         // Adding roles
-        foreach (Role::ALL as $key => $role) {
+        foreach (Role::ALLas $key => $role) {
```
Done:)

viktorcsimma commented on database/migrations/2019_10_19_200553_create_faculties_table.php:
```diff
@@ -18,7 +18,7 @@ public function up()
             $table->text('name');
         });
-        foreach (Faculty::ALL as $key => $faculty) {
+        foreach (Faculty::ALLas $key => $faculty) {
```
Done:)

viktorcsimma commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,488 @@
+            } else {
+                return redirect()->route('reservations.items.show', $item);
+            }
+        }
```
Done:)

viktorcsimma commented on resources/views/livewire/timetable.blade.php:
```diff
@@ -0,0 +1,113 @@
+                                @endphp
+                                @if($isReservation)
+                                <a href="{{ route('reservations.show', App\Models\Reservation::find($block['reservation_id'])) }}">
+                                @elseif(!$isDisabled)
+                                {{-- default values as GET request parameters --}}
+                                <a href="{{ route('reservations.create', ['item' => $item])
+                                            . "?from={$block['from']}&until={$block['until']}"
+                                }}">
+                                @endif
+                                    @php
+                                    if ($isReservation) {
+                                        $reservation = \App\Models\Reservation::find($block['reservation_id']);
+                                        $isOurs = $reservation->user?->is(user());
+                                    } else {
+                                        $reservation = null;
+                                        $isOurs = null;
+                                    }
+                                    @endphp
+                                    <div style="position: absolute;
+                                                left: {{($dayOfWeek * $itemCount + $i) * $columnWidth}}%;
+                                                width: {{$columnWidth}}%;
+                                                top: {{$startHourFloat * 100.0 / 24.0}}%;
+                                                height: {{($endHourFloat - $startHourFloat) * 100.0 / 24.0}}%;"
+                                        @class([
+                                            'timetable-block',
+                                            'valign-wrapper', 'center-align',
+                                            'red' => $isReservation && !$isOurs,
+                                            'orange' => $isOurs,
+                                            'green' => !$isReservation && !$isDisabled,
+                                            'grey' => $isDisabled,
+                                            'darken-4' => $isReservation && App\Models\Reservation::find($block['reservation_id'])->verified
+                                                            || !$isReservation && !$isDisabled,
+                                            'lighten-2' => $isReservation && !App\Models\Reservation::find($block['reservation_id'])->verified
+                                    ])>
+                                        @if(!is_null($reservation))
+                                        {{$reservation->displayName()}}
+                                        @endif
+                                    </div>
```
No, it is important to be able to use PHP variables here.

viktorcsimma commented on database/factories/ReservableItemFactory.php:
```diff
@@ -0,0 +1,29 @@
+            'name' => $this->faker->realText(10),
+            'type' => 'room',
+            'out_of_order' => $this->faker->boolean(30)
+        ];
```
It will be fine for now.

viktorcsimma commented on app/Livewire/Timetable.php:
```diff
@@ -0,0 +1,171 @@
+        $this->lastDay = Carbon::today()->addDays(2);
+
+        $this->calculateBlocks();
+    }
```
Done:)

viktorcsimma commented on app/Mail/ReportReservableItemFault.php:
```diff
@@ -0,0 +1,71 @@
+    {
+        if ($this->item->out_of_order) {
+            return
+                'Javított ' .
+                ($this->isForWashingMachine()
+                 ? 'mosógép' : 'terem')
+                 . ' jelentve';
```
It can be assumed to be read only by admins (who do speak Hungarian).

viktorcsimma committed:
Applying smaller @coderabbitai suggestions

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in c2ed9a0 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/623

viktorcsimma committed:
Simplifying and enhancing UI

viktorcsimma committed:
Fixing erroneous use directive

viktorcsimma committed:
Fixing a minor bug in block logic

viktorcsimma committed:
Rewriting block logic based on OOP and CarbonImmutable instances

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 05da038 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/623

viktorcsimma committed:
Adding tests and mixing bugs discovered by them

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in b0fc6d7 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/623

viktorcsimma committed:
Notifying the secretary and staff on new reservation requests

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in c8bdeb4 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/623

viktorcsimma committed:
Fixing bugs based on tests

viktorcsimma commented on resources/views/emails/reservation_requested.blade.php:
```diff
@@ -0,0 +1,21 @@
+@component('mail::message')
+    @php
+    // formatting date strings
+    $reservedFrom = \Carbon\CarbonImmutable::make($reservation->reserved_from);
+    $reservedUntil = \Carbon\CarbonImmutable::make($reservation->reserved_until);
+    $reservedFrom->settings(['toStringFormat' => 'Y-m-d H:i']);
+    $reservedUntil->settings(['toStringFormat' => 'Y-m-d H:i']);
+    @endphp
```
It's only for here; I'll think it's going to be fine for now.

viktorcsimma committed:
Hiding unverified reservations from those without request rights; adding autoverification for the secretariat and others

viktorcsimma committed:
Fixing a bug related to zero-length reservations

viktorcsimma committed:
Adding notification when someone has rejected or deleted one's reservation

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 7317275 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/623

viktorcsimma committed:
Adding a printable view for room timetables while improving the normal UI

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 82b7452 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/623

viktorcsimma commented on app/Http/Controllers/ReservableItemController.php:
```diff
@@ -0,0 +1,154 @@
+    {
+        $this->authorize('administer', ReservableItem::class);
+
+        $validator = Validator::make($request->all(), [
+            'name' => 'required|string|max:255',
+            'type' => Rule::in(["washing_machine", "room"]),
```
Done:)

viktorcsimma commented on app/Http/Controllers/ReservableItemController.php:
```diff
@@ -0,0 +1,154 @@
+     * (potentially after verifying a fault notice),
+     * or to set it back.
+     */
+    public function toggleOutOfOrder(ReservableItem $item)
+    {
+        $this->authorize('administer', ReservableItem::class);
+
+        $item->update(['out_of_order' => !$item->out_of_order]);
+
+        foreach ($item->usersWithActiveReservation as $toNotify) {
+            Mail::to($toNotify)->send(new AffectedReservation(
+                $item,
+                $toNotify->name
+            ));
+        }
+
+        return redirect()->back()->with('message', __('general.successful_modification'));
+    }
```
Done:)

viktorcsimma commented on app/Http/Controllers/ReservableItemController.php:
```diff
@@ -0,0 +1,154 @@
+     * Sends admins and staff an email notification
+     * about an allegedly faulty item.
+     */
+    public function reportFault(ReservableItem $item)
+    {
+        $this->authorize('requestReservation', $item);
+
+        $thoseToNotify = User::withRole(Role::SYS_ADMIN)->get()
+            ->concat(User::withRole(Role::STAFF)->get());
+        foreach ($thoseToNotify as $toNotify) {
+            Mail::to($toNotify)->send(
+                new ReportReservableItemFault(
+                    $item,
+                    $toNotify->name,
+                    user()->name
+                )
+            );
+        }
+
+        return redirect()->back()->with('message', __('mail.email_sent'));
+    }
```
Done:) I've solved it by queueing.

viktorcsimma commented on app/Http/Controllers/ReservableItemController.php:
```diff
@@ -0,0 +1,154 @@
+    /**
+     * Stores a new item.
+     */
+    public function store(Request $request)
+    {
+        $this->authorize('administer', ReservableItem::class);
+
+        $validator = Validator::make($request->all(), [
+            'name' => 'required|string|max:255',
+            'type' => Rule::in(["washing_machine", "room"]),
+            'out_of_order' => 'nullable|boolean',
+        ]);
+
+        $validatedData = $validator->validate();
+
+        $newItem = ReservableItem::create($validatedData);
+
+        return redirect()->route('reservations.items.show', $newItem);
+    }
```
It will be fine for now.

viktorcsimma commented on app/Http/Controllers/ReservableItemController.php:
```diff
@@ -0,0 +1,154 @@
+    /**
+     * Returns the item creation page.
+     */
+    public function create()
+    {
+        $this->authorize('administer', ReservableItem::class);
+
+        abort(500, 'create not implemented yet');
+    }
```
It is not that important for now.

viktorcsimma commented on app/Http/Controllers/ReservableItemController.php:
```diff
@@ -0,0 +1,154 @@
+     * of the requested type
+     * (or both if there is none given).
+     */
+    public function index(Request $request)
+    {
+        $this->authorize('viewAny', ReservableItem::class);
+
+        $validatedData = $request->validate([
+            'type' => [
+                'required',
+                Rule::enum(ReservableItemType::class)
+            ]
+        ]);
+
+        $items = ReservableItem::where('type', $validatedData['type'])->get();
+        return view('reservations.items.index', [
+            'items' => $items,
+            'type' => $validatedData['type']
+        ]);
+    }
```
I cannot; I depend on the type in the Blade file.

viktorcsimma commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,521 @@
+    /**
+     * Deletes a reservation.
+     */
+    public function deleteAll(Reservation $reservation)
+    {
+        $this->authorize('modify', $reservation);
+
+        if (!$reservation->isRecurring()) {
+            return redirect()->back()->with('error', __('reservations.not_a_recurring_reservation'));
+        }
+
+        // NOTE: this might cause problems if the deletion is unsuccessful
+        // but the message has already got out.
+        if ($reservation->user->id != user()->id) {
+            Mail::to($reservation->user)->queue(new ReservationDeleted(
+                user()->name,
+                $reservation
+            ));
+        }
+
+        $group = $reservation->group;
+        $group->reservations()->delete();
+        $group->delete();
+
+        return redirect()->route('reservations.items.show', $reservation->reservableItem);
+    }
```
Done:) I've reworked it so that it does not use the model itself, but an array.

viktorcsimma commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,521 @@
+    /**
+     * Deletes a reservation.
+     */
+    public function delete(Reservation $reservation)
+    {
+        $this->authorize('modify', $reservation);
+
+        // NOTE: this might cause problems if the deletion is unsuccessful
+        // but the message has already got out.
+        if ($reservation->user->id != user()->id) {
+            Mail::to($reservation->user)->queue(new ReservationDeleted(
+                user()->name,
+                $reservation
+            ));
+        }
+
+        $reservation->delete();
+
+        if ($reservation->reservableItem->isWashingMachine()) {
+            return redirect()->route('reservations.items.index', ['type' => ReservableItemType::WASHING_MACHINE]);
+        } else {
+            return redirect()->route('reservations.items.show', $reservation->reservableItem);
+        }
+    }
```
Same here.

viktorcsimma commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,521 @@
+     * Notifies the secretary and staff on a new or modified reservation that needs to be verified.
+     * Does nothing if the user is a secretary or staff member themselves.
+     */
+    private static function notifyOnVerifiableReservation(Reservation $reservation)
+    {
+        if (!user()->hasRole([Role::SECRETARY, Role::STAFF])) {
+            foreach(User::withRole(Role::SECRETARY)->get()->concat(User::withRole(Role::STAFF)->get())
+                    as $toNotify) {
+                Mail::to($toNotify)->send(
+                    new \App\Mail\ReservationRequested(
+                        $reservation,
+                        $toNotify->name
+                    )
+                );
+            }
+        }
+    }
```
Nice catch:D

viktorcsimma commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,521 @@
+    /**
+     * Checks whether the maximum number of reservations for washing machines has been reached.
+     */
+    private static function reachedMaximumForWashingMachines(User $user): bool
+    {
+        return $user->reservations()
+            ->where('reserved_until', '>', Carbon::now())
+            ->whereHas('reservableItem', function ($query) {
+                $query->where('type', ReservableItemType::WASHING_MACHINE)
+                    ->where('out_of_order', false);
+            })
+            ->count() >= ReservableItem::MAX_WASHING_RESERVATIONS;
+    }
```
I don't understand; isn't your code the same as what I have already written?

viktorcsimma replied:
OK; thank you:)

viktorcsimma commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,521 @@
+    /**
+     * Updates a reservation with an edited version.
+     */
+    public function update(Reservation $reservation, Request $request)
+    {
+        $this->authorize('modify', $reservation);
+
+        $item = $reservation->reservableItem;
+
+        if ($item->isOutOfOrder()) {
+            return redirect()->back()->with('error', __('reservations.item_out_of_order'));
+        } elseif (Carbon::make($reservation->reserved_until) < Carbon::now()) {
+            return redirect()->back()->with('error', __('reservations.editing_past_reservations'));
+        }
```
I moved that logic into the 'modify' policy; thanks!

viktorcsimma commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,521 @@
+        } elseif ($item->isWashingMachine()
+            && self::reachedMaximumForWashingMachines(user())) {
+            return redirect()->back()->with('error', __('reservations.max_washing_reservations_reached'));
+        }
```
I've moved the first part to a policy; the second would not be worth a separate function IMO.

viktorcsimma commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,521 @@
+            fn ($input) => $item->isWashingMachine()
+        );
+
+        if ($item->isWashingMachine()) {
+            $validator->after(function ($validator) use ($request) {
+                $from = Carbon::make($request->reserved_from);
+                // we only allow one-hour slots for washing machines
+                if (0 != $from->minute || 0 != $from->second) {
+                    $validator->errors()->add('reserved_from', __('reservations.one_hour_slot_only'));
+                }
+            });
+        }
```
I think this is not worth a separate function... it is not a meaningful code block by itself.

viktorcsimma committed:
Applying @coderabbitai suggestions

viktorcsimma committed:
Merge branch 'development' into reservable-items-final

viktorcsimma committed:
Fixing illegal calls

viktorcsimma committed:
Adding recurring reservations to the seeder

viktorcsimma committed:
Fixing some things on the print view

viktorcsimma commented on app/Mail/ReservationDeleted.php:
```diff
@@ -0,0 +1,91 @@
+     *
+     * @return \Illuminate\Mail\Mailables\Envelope
+     */
+    public function envelope()
+    {
+        return new Envelope(
+            subject:
+                $this->reservationArray['verified']
+                ? __('reservations.reservation_deleted')
+                : __('reservations.reservation_rejected'),
+        );
```
Nice; I'll take this one:)

viktorcsimma commented on app/Models/User.php:
```diff
@@ -461,6 +461,22 @@ public function presenceChecks(): BelongsToMany
+    /**
+     * @return BelongsToMany The reservable items which the user has ever reserved.
+     */
+    public function reservableItems(): BelongsToMany
+    {
+        return $this->belongsToMany(ReservableItem::class, Reservation::class, 'user_id', 'reservable_item_id');
+    }
```
We might do so.

viktorcsimma committed:
Applying @coderabbitai suggestions

viktorcsimma committed:
Adding a .env variable for closing room reservation except for the secretariat

viktorcsimma committed:
Switching from `random_int` to `rand` in seeders

viktorcsimma committed:
Adjusting the timetable to mobile screens too

viktorcsimma commented on database/factories/EducationalInformationFactory.php:
```diff
@@ -23,7 +23,7 @@ public function definition()
             'year_of_acceptance' => $this->faker->numberBetween($min = 2015, $max = date('Y')),
             'email' => $this->faker->unique()->safeEmail,
-            'research_topics' => $this->faker->paragraph(random_int(1, 3)),
+            'research_topics' => $this->faker->paragraph(rand(1, 3)),
```
This is a factory; we don't need a CSPRNG here.

viktorcsimma committed:
Applying most suggestions

viktorcsimma committed:
Adding a date picker above the timetable to choose the first day

viktorcsimma committed:
Forbidding to edit reservations that have already begun

BertalanD committed:
Move timetable date picker in line with prev/next day buttons

viktorcsimma committed:
Clarifying use of the datepicker

viktorcsimma committed:
Hiding reservation instructions from those unauthorized to request reservations

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 0a20d00 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/623

viktorcsimma committed:
Fixing a test case

kdmnk commented:
Hi! Sorry, I can take a quick look again tomorrow morning. I know you want to merge soon so we'll see

viktorcsimma committed:
Hiding room reservations from tenants

viktorcsimma commented:
@kdmnk thank you:)

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 39f28bb according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/623

viktorcsimma committed:
Fixing bugs

viktorcsimma commented on app/Http/Controllers/ReservableItemController.php:
```diff
@@ -0,0 +1,159 @@
+    /**
+     * Returns the item creation page.
+     */
+    public function create()
+    {
+        $this->authorize('administer', ReservableItem::class);
+
+        abort(500, 'create not implemented yet');
+    }
```
This will be fine for now.

viktorcsimma committed:
Adding a receptionist role

viktorcsimma committed:
Fixing bugs related to recurring reservations stretching through midnight

kdmnk commented on app/Exceptions/ConflictException.php:
```diff
@@ -0,0 +1,8 @@
+namespace App\Exceptions;
+
+/** Thrown in transactions when there is a conflict somewhere. */
+class ConflictException extends \Exception
```
rename to resevation conflict exception

kdmnk commented on app/Policies/ReservableItemPolicy.php:
```diff
@@ -0,0 +1,110 @@
+     * (not counting if it is out of order etc.).
+     * Note: `canViewType` calls this too.
+     */
+    public static function canRequestReservationForType(User $user, ReservableItemType $type): bool
```
you can make these non static and call them in the conventional way: authorize('canRequestReservationForType', [ReservableItem::class, $type]).

kdmnk commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,506 @@
+     * a ReservableItem provided separately
+     * and the data in the request.
+     */
+    public function store(ReservableItem $item, Request $request)
```
Split this into multiple functions. For example (from chatgpt):

   ```
 $this->authorize('requestReservation', $item);

    if ($item->isWashingMachine() && self::reachedMaximumForWashingMachines(user())) {
        return redirect()->back()->with('error', __('reservations.max_washing_reservations_reached'));
    }

    $validatedData = $this->validateReservation($request, $item);

    if ($item->isRoom() && isset($request->recurring)) {
        return $this->handleRecurringReservation($validatedData, $item, $request);
    } else {
        return $this->handleSingleReservation($validatedData, $item);
    }
```

kdmnk commented on app/Http/Controllers/ReservationController.php:
```diff
@@ -0,0 +1,506 @@
+    /**
+     * Updates a reservation with an edited version.
+     */
+    public function update(Reservation $reservation, Request $request)
```
Same, split this into multiple functions. If makes sense, you could create an interface and two more classes for handling washing machines and for rooms.

kdmnk commented on app/Models/ReservationGroup.php:
```diff
@@ -0,0 +1,336 @@
+            $allAfter = $this->reservations()
+                ->where('reserved_from', '>=', $firstReservation->reserved_from)
+                ->get();
+            foreach($allAfter as $reservation) {
```
foreach body is too long, should be a new private function

kdmnk reviewed (approved):
This is a crazy huge module, if everything works, good job!
Despite the approval, I left some comments, I hope you can take a look.

And one more thing: please put all related Models, Emails and Email blades into its own subdirectory as those are starting to be too many in a flat hierarchy.

Understanding the code might take some time, but let's see how many times do you have to come back here to fix or extend some functionality... Otherwise, it is quite separated from the rest of the system so I think it is fine.
I recommend to come back here some time later and try to improve the software architecture. It's not trivial but a nice example for improving that.

And about the receptionists: we had some issues some years ago with the fault reporting, that they do not have a computer there. I don't know if it is still a problem, but better to take a look and ensure that it is feasible for them to access the system. Otherwise, you could put a screen in the entrance hall showing the current status of the rooms. 

Good job again, it is a long awaited module, I hope it will be successful!

viktorcsimma committed:
Restructuring file hierarchy

viktorcsimma committed:
Normalizing strange policies

viktorcsimma committed:
Restructuring reservation-related mailables into folders

viktorcsimma committed:
Refactoring ReservationController; fixing bugs

viktorcsimma committed:
Moving one more Mailable to the correct directory

viktorcsimma committed:
Checking mails in tests

viktorcsimma committed:
Clarifying UI in parts related to recurring events

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 0ed4eb8 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/623

viktorcsimma committed:
Merge branch 'development' into reservable-items-final

viktorcsimma committed:
Refactoring the core of a foreach loop into a separate function

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 70bbf0f according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/623

viktorcsimma committed:
Fixing an erroneous type hint

viktorcsimma commented:
Thank you so much:) I have applied almost all of the things you have suggested. I have not tried interfaces for ReservationController though; I think I will do so when redesigning the architecture later (which I should).

What do you think, can we proceed with the merge? I would try it on staging then.

viktorcsimma committed:
Adding a missing .blade.php extension

viktorcsimma commented on resources/views/emails/reservations/reservation_affected.php:
```diff
@@ -0,0 +1,20 @@
+@component('mail::message')
+    <h1>@lang('mail.dear') {{ $recipient }}!</h1>
```
Oh; I've forgotten the `.blade` from the name. Thanks:)

viktorcsimma commented on app/Http/Controllers/Dormitory/Reservations/ReservationController.php:
```diff
@@ -0,0 +1,585 @@
+    /**
+     * Deletes a reservation.
+     */
+    public function deleteAll(Reservation $reservation)
+    {
+        $this->authorize('modify', $reservation);
+
+        if (!$reservation->isRecurring()) {
+            return redirect()->back()->with('error', __('reservations.not_a_recurring_reservation'));
+        }
+
+        // we will need these for the mailable
+        $ownerName = $reservation->user->name;
+        $itemName = $reservation->reservableItem->name;
+        $reservationArray = $reservation->toArray();
+
+        $group = $reservation->group;
+        $group->reservations()->delete();
+        $group->delete();
+
+        if ($reservation->user->id != user()->id) {
+            Mail::to($reservation->user)->queue(new ReservationDeleted(
+                user()->name,
+                $ownerName,
+                $itemName,
+                $reservationArray,
+                isForAll: true
+            ));
+        }
+
+        return redirect()->route('reservations.items.show', $reservation->reservableItem);
```
Done:)

viktorcsimma commented on app/Http/Controllers/Dormitory/Reservations/ReservationController.php:
```diff
@@ -0,0 +1,585 @@
+        return $validator->validate();
+    }
+
+    /**
+     * An auxiliary function for `store`.
+     * Handles validated data which we know to belong to
+     * a new recurring reservation.
+     * Returns the first reservation of the new group.
+     * If there has been a conflict,
+     * it throws a ReservationConflictException
+     * with the error message to show to the user
+     * and does nothing to the database.
+     */
+    private static function handleNewRecurringReservation(ReservableItem $item, array $validatedData): Reservation
+    {
+        $newGroup = ReservationGroup::create([
+            'group_item' => $item->id,
+            'user_id' => user()->id,
+            'frequency' => intval($validatedData['frequency']),
+            'group_title' => $validatedData['title'],
+            'group_from' => Carbon::make($validatedData['reserved_from']),
+            'group_until' => Carbon::make($validatedData['reserved_until']),
+            'group_note' => $validatedData['note'],
+            'last_day' => Carbon::make($validatedData['last_day']),
+            'verified' => user()->can('autoVerify', $item)
+        ]);
+
+        try {
+            $newGroup->initializeFrom($validatedData['reserved_from']);
+        } catch (ReservationConflictException $e) {
+            $newGroup->delete();
+            throw $e;
+        }
+
+        $firstReservation = $newGroup->firstReservation();
+        return $firstReservation;
+    }
```
Done:)

viktorcsimma commented on app/Mail/Reservations/ReportReservableItemFault.php:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * The message subject.
+     */
+    public function makeSubject(): string
+    {
+        if ($this->item->out_of_order) {
+            return
+                'Javított ' .
+                ($this->isForWashingMachine()
+                 ? 'mosógép' : 'terem')
+                 . ' jelentve';
+        } else {
+            return
+                $this->isForWashingMachine()
+                ? 'Hibás mosógép jelentve'
+                : 'Használhatatlan terem jelentve';
+        }
+    }
```
Done:)

viktorcsimma commented on app/Http/Controllers/Dormitory/Reservations/ReservableItemController.php:
```diff
@@ -0,0 +1,158 @@
+    {
+        $this->authorize('view', $item);
+
+        $thoseToNotify = User::withRole(Role::SYS_ADMIN)->get()
+            ->concat(User::withRole(Role::STAFF)->get());
+        foreach ($thoseToNotify as $toNotify) {
+            Mail::to($toNotify)->queue(
+                new ReportReservableItemFault(
+                    $item,
+                    $toNotify->name,
+                    user()->name
+                )
+            );
+        }
```
Done:)

viktorcsimma committed:
Applying some CodeRabbit suggestions

viktorcsimma committed:
Adjusting the `show` and `edit` pages better to washing machine reservations
